캠 뷰 컴포넌트 제작

voidmatcha opened the pull request:
> resolve #32 #46 #67
## #32
- [x] video 태그로 웹캠 소스를 재생
- [x] 이름 라벨
- [x] REC 라벨
- [x] 버튼: 재생, 캠, 마이크
- [x] 1:1 연습시 하단 상대방캠 활성화
## #46
- [x] 기본적인 녹화 세팅 완료 및 useMediaRecorder custom hooks 추가
- [x] Kafka와 연동할 수 있도록 blob 세팅 - application/octet-stream

https://yapp-17th.github.io/Web_2_Client/feat/32/?path=/story/camview--normal-cam-view

voidmatcha committed:
fix: .DS_Store를 .gitignore에 추가

voidmatcha committed:
feat: 캠 뷰 컴포넌트 - 캠, 이름, REC 라벨 컴포넌트 제작 완료 (#32)
- 캠 화면 컴포넌트 제작 완료
- 이름 라벨 컴포넌트 제작 완료
- REC 라벨 컴포넌트 제작 완료
해야 할 일
- [ ] 버튼: 재생, 캠, 마이크
- [ ] 하단 상대방캠 활성화

SkynI25 commented:
와..라벨 붙이셨군요.. 👍 저도 작업할때 참고하겠습니다 ㅎㅎ

voidmatcha committed:
feat(record): Web API 를 이용해 현재 탭 및 스테레오 믹스 녹화하는 custom hooks 추가 (#46)
- 기본적인 녹화 세팅 완료 및 useMediaRecorder custom hooks 추가
- [ ] kafka 연동 세팅 필요 - application/octet-stream

voidmatcha committed:
feat: 캠 뷰 컴포넌트 - 하단 상대방캠 활성화 제작 완료 (#32)
- oneOnOne props가 true이면 peerRef를 이용해 상대방 캠 활성화 됨

Lavegaa commented:
오우 고생 많으셨습니다. 확인해보겠습니다.

voidmatcha committed:
Merge branch 'master' into feat/#32

voidmatcha committed:
feat: useMediaRecorder hook에 서버전송 위한 blob 생성/1s 코드 추가 (#46)
- mediaRecorder.current.requestData() 와 mediaChunkEach.current 를 setInterval을 활용해 구현
- 1초마다 생성되는 blob property에 application/octet-stream 설정
- 전체 녹화된 blob property는 vidoe/mp4 설정
- 1초마다 생성되는 blob, 전체 녹화 blob 둘다 가져올 수 있음
- lint에 맞게 코드 수정

voidmatcha committed:
fix(record): 전체 녹화 blob이 작동하지 않는 문제 해결
- stop시 mediaChunks, mediaChunkEach 초기화 위치가 문제였음

voidmatcha committed:
refactor: smallCamView 컴포넌트 PropTypes 추가
- videoRef는 object 형식이라고 함

voidmatcha committed:
refactor: CamView Storybook 코드 스타일 리펙터
- useReactMediaRecorder hook이 default로 export 하기 때문에 분해할당 하지 않아도 됨
- 코드 정리

voidmatcha committed:
refcator: CamView 하위 컴포넌트들 코드 정리 (#46)
- 코드 정리

voidmatcha commented on src/hooks/useMediaRecorder.js:
```diff
@@ -0,0 +1,218 @@
+  const [mediaBlobUrl, setMediaBlobUrl] = useState(null);
+  const [error, setError] = useState('NONE');
+
+  useEffect(() => {
+    if (status === 'recording') {
+      setInterval(() => {
+        mediaRecorder.current.requestData();
+
+        const blobProperty = { type: 'application/octet-stream' };
+
+        const blobEach = new Blob([mediaChunkEach.current.slice(-1)[0]], blobProperty);
+
+        // TODO: console.log 지우고 서버(kafka)에 업로드하는 코드 추가해야 함
+        console.log(blobEach);
+      }, 1000);
+    } else {
+      for (let i = 1; i < 99999; i += 1) window.clearInterval(i);
+    }
+  }, [status]);
```
이 코드가 1초마다 생성되는 blob을 서버로 보내기 위한 코드입니다.

voidmatcha committed:
Merge branch 'master' into feat/#32

voidmatcha commented on src/hooks/useMediaRecorder.js:
```diff
@@ -0,0 +1,218 @@
+        // TODO: console.log 지우고 서버(kafka)에 업로드하는 코드 추가해야 함
+        console.log(blobEach);
+      }, 1000);
+    } else {
+      for (let i = 1; i < 99999; i += 1) window.clearInterval(i);
+    }
```
혹시 clearInterval 좀  더 깔끔하게 하는 방법이 있을까요?

SkynI25 replied:
당장 떠오르는 방법은 없지만 한번 찾아서 공유해드릴게요!

Lavegaa replied:
https://stackoverflow.com/questions/34167975/clear-all-setintervals/34168038
interval를 저장하고 foreach로 지우는 경우가 있긴 하네요
```js
var intervals = [];
$(".elements").each(function() {
    var i = setInterval(function() {

    }, 1000);
    intervals.push(i);
});

...

intervals.forEach(clearInterval);
```
해당 방법으로 가능할까요 흠

voidmatcha replied:
한번 적용해보도록 하겠습니다!

voidmatcha replied:
잘 작동하네요 적용했습니다!

voidmatcha committed:
chore: package-lock.json 업데이트
- master에 merge되고 새로 설치되어야할 모듈이 있어서 설치함

voidmatcha committed:
fix: Icon 컴포넌트 원 아이콘 그림자에 대해 border-radius 옵션 추가 (#67)
- 정사각혁으로 그람지가 짤리기 때문에 어색함
- Icon 컴포넌트에 props 옵션 추가

voidmatcha committed:
feat: 1:1 연습시 캠 하단 버튼 그룹 완성 (#32)
- Redux로 useMediaRecorder hook과 상태 연동해서 작동할 수 있게 해야 함
- 버튼이 눌렸을 때의 아이콘이 필요함

SkynI25 commented on src/components/CamView/ButtonGroup.js:
```diff
@@ -0,0 +1,43 @@
+  }
+`;
+export default function ButtonGroup() {
+  // TODO: redux 붙여서 상태 관리될 것
```
이 부분 주석은 어떤 것을 의미하는 걸까요? ㅎㅎ

voidmatcha replied:
이부분은 버튼 Icon 이 눌렸을때를 redux로 상태 관래서 구분해야한다는 의도로 넣어놓은 주석입니다.

SkynI25 commented on src/components/CamView/ButtonGroup.js:
```diff
@@ -0,0 +1,43 @@
+  return (
+    <Wrapper>
+      <WrapIcon>
+        <Icon type="play_blue" isCircle alt="play_blue" />
```
Icon 컴포넌트에 isCircle 이란 prop 을 추가하셨네요. 해당 prop을 주면 좀 더 원에 가까운 형태로 표현할 수 있나요?

voidmatcha replied:
아니요 기태님이 겼으신것과 같이 #67 원형아이콘인데 그람자같은 배경효과가 있는 경우 사각형으로 어색한 상황이 있을 때 이를 제거하는 props입니다. https://github.com/Yapp-17th/Web_2_Client/pull/45/commits/f493e44919bc90ccdd0469e99453bce3a5aee933 이 커밋을 cherry-pick해서 사용하시면 됩니다!

SkynI25 commented on src/components/CamView/CamView.js:
```diff
@@ -0,0 +1,105 @@
+    setupStream();
+  }, [peerRef]);
+
+  // TODO: 해당 useEffect는 잘 녹화되었는지 테스트 목적 - 실제 서비스에선 제거되어야 함
```
향후 진행할 작업에 관해서도 주석을 깔끔하게 넣으셨네요, 저도 메모할겸 이런 방식으로 코드 주석을 남겨봐야 겠습니다!

SkynI25 reviewed (approved):
정말 수고많으셨습니다.. 영상 처리쪽을 진행하시면서 기술적 어려움도 많으셨을텐데.. 전체적으로 깔끔하게 코드를 구현하신 게 인상적이네요! :) 저도 영상 처리를 구현할 때 많이 참고하겠습니다 ㅎㅎ

Lavegaa commented on src/hooks/useMediaRecorder.js:
```diff
@@ -0,0 +1,218 @@
+  const getMediaStream = useCallback(async () => {
+    setStatus('acquiring_media');
+    const requiredMedia = {
+      audio: typeof audio === 'boolean' ? !!audio : audio,
```
!!로 확실하게 falsy, truthy한 표현이 가능하군요! 좋은 지식 배워갑니다.

Lavegaa commented on src/hooks/useMediaRecorder.js:
```diff
@@ -0,0 +1,218 @@
+        );
+        mediaStream.current = stream;
+      }
+      setStatus('idle');
```
녹화에서 record, stop만 생각했는데 idle도 챙겼어야 하네요!

Lavegaa reviewed (approved):
영상쪽에 대해서는 스터디좀 많이 해야겠네요... 대단하십니다! 고생 많으셨어요!!

voidmatcha committed:
refactor: setInterval의 clearInterval 방식 개선 (#45)
- 기존에 99999번까지 모든 interval을 clear 하는 방식에서 useState에 값을 넣어 해당하는 것만 clear 하는 방식으로 변경
